test/ehttpc_server.erl

```diff
@@ -91,42 +84,101 @@ listen(Port) ->
     end.
 
 loop(Socket, Opts) ->
-    loop(Socket, Opts, <<>>).
+    loop(fetch_header, Socket, Opts, <<>>).
 
-loop(Socket, Opts, Buffer) ->
-    Delay0 = maps:get(delay, Opts, 0),
+loop(check_header_and_continue_fetch_if_not_ready, Socket, Opts, Buffer) ->
+    case string:find(Buffer, <<"\r\n\r\n">>) of
+        nomatch ->
+            loop(fetch_header, Socket, Opts, Buffer);
+        _Match ->
+            loop(parse_header, Socket, Opts, Buffer)
+    end;
+loop(fetch_header, Socket, Opts, Buffer) ->
+    %% Read data until we get a double new line
+    %% This means that we have received the header and can go on to parse the header
     case gen_tcp:recv(Socket, 0) of
-        {ok, Data} ->
-            Split = binary:split(
-                <<Buffer/binary, Data/binary>>,
-                <<"\r\n\r\n">>,
-                [global]
-            ),
-            {N, Buffer2} = count_requests(Split, 0),
-            Responses = make_responses(N, Opts),
-            Ms =
-                case Delay0 of
-                    {rand, D} -> rand:uniform(D);
-                    _ -> Delay0
-                end,
-            Ms > 0 andalso delay(Socket, Ms),
-            case socket_send(Socket, Responses, Opts) of
-                ok ->
-                    case maps:get(oneoff, Opts, false) of
-                        true ->
-                            gen_tcp:shutdown(Socket, write),
-                            exit(normal);
-                        false ->
-                            ok
-                    end,
-                    loop(Socket, Opts, Buffer2);
-                {error, _Reason} ->
+        {ok, Data0} ->
+            Data1 = <<Buffer/binary, Data0/binary>>,
+            loop(check_header_and_continue_fetch_if_not_ready, Socket, Opts, Data1);
+        {error, _Reason} ->
```

Comment:
should probably match on a normal error e.g. `tcp_closed` ? or at least add a comment here.
```suggestion
        {error, _Reason} ->
            %% socket closed ? stop looping
            ok
```